Validation/HGCalValidation/plugins/CaloParticleValidation.cc

```diff
@@ -0,0 +1,267 @@
+  caloParticles_(consumes<std::vector<CaloParticle> >(iConfig.getParameter<edm::InputTag>("caloParticles"))),
+  simPFClusters_(consumes<std::vector<reco::SuperCluster>>(iConfig.getParameter<edm::InputTag>("simPFClusters"))),
+  simPFCandidates_(consumes<reco::PFCandidateCollection>(iConfig.getParameter<edm::InputTag>("simPFCandidates"))),
+  recHitsEE_(consumes<HGCRecHitCollection>(iConfig.getParameter<edm::InputTag>("recHitsEE"))),
+  recHitsFH_(consumes<HGCRecHitCollection>(iConfig.getParameter<edm::InputTag>("recHitsFH"))),
+  recHitsBH_(consumes<HGCRecHitCollection>(iConfig.getParameter<edm::InputTag>("recHitsBH")))
+{
+   //now do what ever initialization is needed
+}
+
+
+CaloParticleValidation::~CaloParticleValidation()
+{
+   // do anything here that needs to be done at desctruction time
+   // (e.g. close files, deallocate resources etc.)
+}
+
+
+//
+// member functions
+//
+
+// ------------ method called for each event  ------------
+
+void
+CaloParticleValidation::dqmAnalyze(edm::Event const& iEvent, edm::EventSetup const& iSetup,
+                      Histograms_CaloParticleValidation const & histos) const
+{
+  using namespace edm;
+
+  Handle<HGCRecHitCollection> recHitHandleEE;
+  Handle<HGCRecHitCollection> recHitHandleFH;
+  Handle<HGCRecHitCollection> recHitHandleBH;
+  // make a map detid-rechit
+
+  iEvent.getByToken(recHitsEE_, recHitHandleEE);
+  iEvent.getByToken(recHitsFH_, recHitHandleFH);
+  iEvent.getByToken(recHitsBH_, recHitHandleBH);
+  const auto& rechitsEE = *recHitHandleEE;
+  const auto& rechitsFH = *recHitHandleFH;
+  const auto& rechitsBH = *recHitHandleBH;
+  std::map<DetId, const HGCRecHit*> hitmap;
+  for (unsigned int i = 0; i < rechitsEE.size(); ++i) {
+    hitmap[rechitsEE[i].detid()] = &rechitsEE[i];
+  }
+  for (unsigned int i = 0; i < rechitsFH.size(); ++i) {
+    hitmap[rechitsFH[i].detid()] = &rechitsFH[i];
+  }
+  for (unsigned int i = 0; i < rechitsBH.size(); ++i) {
+    hitmap[rechitsBH[i].detid()] = &rechitsBH[i];
+  }
+
+  Handle<std::vector<SimVertex>> simVerticesHandle;
+  iEvent.getByToken(simVertices_, simVerticesHandle);
+  std::vector<SimVertex> const & simVertices = *simVerticesHandle;
+
+  Handle<std::vector<CaloParticle> > caloParticleHandle;
+  iEvent.getByToken(caloParticles_, caloParticleHandle);
+  std::vector<CaloParticle> const & caloParticles = *caloParticleHandle;
+
+  Handle<std::vector<reco::SuperCluster>> simPFClustersHandle;
+  iEvent.getByToken(simPFClusters_, simPFClustersHandle);
+  std::vector<reco::SuperCluster> const & simPFClusters = *simPFClustersHandle;
+
+  Handle<reco::PFCandidateCollection> simPFCandidatesHandle;
+  iEvent.getByToken(simPFCandidates_, simPFCandidatesHandle);
+  reco::PFCandidateCollection const & simPFCandidates = *simPFCandidatesHandle;
+
+  for (auto const caloParticle : caloParticles) {
+    int id = caloParticle.pdgId();
+    if (histos.count(id)) {
+      histos.at(id).eta_.fill(caloParticle.eta());
```

Comment:
maybe faster to do `auto& histo = histos.at(id);` once rather than calling `histos.at(id)` a bunch of times?